packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/subscribe.ts

```diff
@@ -0,0 +1,294 @@
+   */
+  readonly subscriptionQueue: sqs.IQueue;
+}
+
+/**
+ * Settings for the hook which mutates the application container
+ * to add the subscription queue URLs to its environment.
+ */
+export interface ContainerMutatingProps {
+  /**
+   * The queue name and URLs to be added to the container environment.
+   */
+  readonly environment: { [key: string]: string };
+}
+
+/**
+ * This hook modifies the application container's environment to
+ * add the queue URLs for the subscription queues of the service.
+ */
+export class SubscribeMutatingHook extends ContainerMutatingHook {
+  private environment: { [key: string]: string };
+
+  constructor(props: ContainerMutatingProps) {
+    super();
+    this.environment = props.environment;
+  }
+
+  public mutateContainerDefinition(props: ecs.ContainerDefinitionOptions): ecs.ContainerDefinitionOptions {
+    return {
+      ...props,
+
+      environment: { ...(props.environment || {}), ...this.environment },
+    } as ecs.ContainerDefinitionOptions;
+  }
+}
+
+/**
+ * This extension creates a default SQS Queue, `eventsQueue`. It also supports creation of SQS Queue
+ * per SNS Topic and sets up the SNS Subscriptions accordingly.
+ *
+ * The created subscriptions can be accessed using `subscriptions` field and the
+ * default queue for this service can be accessed using the getter `<extension>.eventsQueue`.
+ */
+export class SubscribeExtension extends ServiceExtension {
+  /** The subscriptions created by the extension.
+   * @default undefined
+   */
+  public readonly subscriptions?: Subscription[];
+
+  private _eventsQueue!: sqs.IQueue;
+
+  private environment: { [key: string]: string } = {};
+
+  private deadLetterQueues: sqs.IQueue[] = [];
+
+  private props?: SubscribeProps;
+
+  constructor(props?: SubscribeProps) {
+    super('subscribe');
+
+    this.props = props;
+    if (props?.topicSubscriptions) {
+      this.subscriptions = [];
+    }
+
+    if (props?.eventsQueue && props?.eventsQueueProps) {
+      throw Error('You can only specify either a queue or queue props for creating the default events queue.');
+    }
+  }
+
+  /**
+   * This hook creates and sets the default queue `eventsQueue` and dead letter queue and other topic-specific
+   * according to the provided settings. It creates subscriptions for the topics and also adds the queue URLs to the
+   * environment.
+   * @param service The parent service which this extension has been added to
+   * @param scope The scope that this extension should create resources in
+   */
+  public prehook(service: Service, scope: Construct) {
+    this.parentService = service;
+    this.scope = scope;
+
+    if (this.props?.eventsQueue) {
+      this._eventsQueue = this.props.eventsQueue;
+    } else {
+      this._eventsQueue = this.createQueues('Events', this.props?.eventsQueueProps, this.props?.eventsDeadLetterQueueProps);
+    }
+    this.environment[`${this.parentService.id.toUpperCase()}_EVENTS_QUEUE_URL`] = this._eventsQueue.queueUrl;
```

Comment:
Can we standardize these variable names with Copilot? 

For example, we'll use
```
COPILOT_QUEUE_URI for the primary queue
COPILOT_TOPIC_QUEUE_URIS for a JSON map: {"service-topic": "https://queue-uri.amazonaws.com"}

COPILOT_SNS_TOPIC_ARNS
```

Could we do something similar and swap COPILOT for ${PARENTSERVICE}?

Reply:
Actually, the environment in CDK is typed in to be a set of key-value pairs of strings. That sort of limits our ability to add the variables exactly like that. So I have added an object called `subscriptions` that is a mapping of topics and their subscribed queue and will serve the same purpose if the user needs to refer to it later.

Reply:
Update: As the topic-specific queues will be user input, we will be removing the addition of these  queueURLs as environment variables. The only queue possibly created by us is the default `eventsQueue`, and that URL will still be added to the environment.